Regression: Snapshot on CIFS fails, fallback to move from hardlink not working

### Elasticsearch Version

8.19.16

### Installed Plugins

xpack

### Java Version

_bundled_

### OS Version

Debian GNU/Linux 12 (bookworm)

### Problem Description

After updating a cluster to elasticsearch 8.19.16 (from 8.19.15), creating snapshots on a specific CIFS share stopped working.

The root cause is: `java.nio.file.AccessDeniedException: /mnt/backups/index-1821 -> /mnt/backups/pending-index-1821-rSDrglkWTdGSJn4r1YBJzw`, thrown in `java.base/java.nio.file.Files.createLink(Files.java:1020)` (see full log below).

We have narrowed it down to not being able to create hardlinks on this specific share (other clusters with different shares are able to snapshot correctly). This behaviour was introduced in 8.19.16 ([PR 147405](https://github.com/elastic/elasticsearch/pull/147405), [PR 148777](https://github.com/elastic/elasticsearch/pull/148777), #148811) - the exception handling seems to be insufficient under some circumstances, as our share throws a `AccessDeniedException`, not the properly handled `FileSystemException`.

### Steps to Reproduce

We cannot provide a guide to reproduction, as we cannot provide the configuration of the specific share.

The share does not allow hard links - we verified that.

### Logs (if relevant)

```shell
[2026-06-18T03:32:48,643][WARN ][o.e.s.SnapshotsService   ] [myhostname] [backup:daily-snap-2026.06.18-bnd_ciiws-ga7yaq8a4pdw/bvJKnMKkT3qLLF5DgBPWtg] failed to finalize snapshot
org.elasticsearch.snapshots.SnapshotException: [backup:daily-snap-2026.06.18-bnd_ciiws-ga7yaq8a4pdw/bvJKnMKkT3qLLF5DgBPWtg] failed to update snapshot in repository
        at org.elasticsearch.server@8.19.16/org.elasticsearch.repositories.blobstore.BlobStoreRepository.lambda$finalizeSnapshot$23(BlobStoreRepository.java:1891)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.acceptException(ActionListenerImplementations.java:188)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:64)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onFailure(ActionListenerImplementations.java:193)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.support.SubscribableListener$FailureResult.complete(SubscribableListener.java:408)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.support.SubscribableListener.tryComplete(SubscribableListener.java:320)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.support.SubscribableListener.setResult(SubscribableListener.java:345)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.support.SubscribableListener.onFailure(SubscribableListener.java:264)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:64)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:75)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.support.SubscribableListener$FailureResult.complete(SubscribableListener.java:408)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.support.SubscribableListener.tryComplete(SubscribableListener.java:320)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.support.SubscribableListener.setResult(SubscribableListener.java:345)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.support.SubscribableListener.onFailure(SubscribableListener.java:264)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:64)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:75)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:64)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:75)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:249)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.support.SubscribableListener$SuccessResult.complete(SubscribableListener.java:396)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.support.SubscribableListener.tryComplete(SubscribableListener.java:316)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.support.SubscribableListener.setResult(SubscribableListener.java:345)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.support.SubscribableListener.onResponse(SubscribableListener.java:259)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.repositories.blobstore.BlobStoreRepository.lambda$writeIndexGen$39(BlobStoreRepository.java:2847)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.ActionRunnable$4.doRun(ActionRunnable.java:101)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1044)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
        at java.base/java.lang.Thread.run(Thread.java:1516)
Caused by: java.nio.file.AccessDeniedException: /mnt/backups/index-1821 -> /mnt/backups/pending-index-1821-rSDrglkWTdGSJn4r1YBJzw
        at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:90)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:108)
        at java.base/sun.nio.fs.UnixFileSystemProvider.createLink(UnixFileSystemProvider.java:528)
        at java.base/java.nio.file.Files.createLink(Files.java:1020)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.common.blobstore.fs.FsBlobContainer.moveAtomicallyUsingHardLink(FsBlobContainer.java:377)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.common.blobstore.fs.FsBlobContainer.moveBlobAtomic(FsBlobContainer.java:355)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.common.blobstore.fs.FsBlobContainer.writeMetadataBlob(FsBlobContainer.java:268)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.repositories.blobstore.BlobStoreRepository.writeAtomic(BlobStoreRepository.java:3183)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.repositories.blobstore.BlobStoreRepository.lambda$writeIndexGen$42(BlobStoreRepository.java:2870)
        at org.elasticsearch.server@8.19.16/org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:247)
        ... 11 more
[2026-06-18T03:32:48,685][ERROR][o.e.x.s.SnapshotLifecycleTask] [myhostname] failed to create snapshot for snapshot lifecycle policy [daily-snapshots]: org.elasticsearch.snapshots.SnapshotException: [backup:daily-snap-2026.06.18-bnd_ciiws-ga7yaq8a4pdw/bvJKnMKkT3qLLF5DgBPWtg] failed to update snapshot in repository
```

--- a/docs/changelog/153219.yaml
+++ b/docs/changelog/153219.yaml
@@ -1,0 +1,6 @@
+area: Snapshot/Restore
+issues:
+ - 152053
+pr: 153219
+summary: Fix snapshot creation on CIFS shares in case of access denied exception
+type: bug

--- a/server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java
+++ b/server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java
@@ -434,9 +434,10 @@ public class FsBlobContainer extends AbstractBlobContainer {
             return;
         } catch (FileSystemException e) {
             // In some filesystems (e.g., CIFS) the Files.createLink() does not throw UnsupportedOperationException (as Javadoc says) but
-            // FileSystemException instead. To handle this, we fall back to the no-hard-link operation. This is an indirect check
-            // that assumes that direct FileSystemExceptions instance creation is rare, and it indicates a lack of support for hard links.
-            if (e.getClass() != FileSystemException.class) {
+            // FileSystemException or AccessDeniedException instead. To handle this, we fall back to the no-hard-link operation. This is an
+            // indirect check that assumes that direct FileSystemExceptions or AccessDeniedException instances are rare, and they indicate
+            // a lack of support for hard links or insufficient permissions for creating them.
+            if (e.getClass() != FileSystemException.class && e instanceof AccessDeniedException == false) {
                 throw e;
             }
             fallbackMoveAtomically.accept(targetBlobPath, sourceBlobPath);

--- a/server/src/test/java/org/elasticsearch/common/blobstore/fs/FsBlobContainerTests.java
+++ b/server/src/test/java/org/elasticsearch/common/blobstore/fs/FsBlobContainerTests.java
@@ -31,6 +31,7 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.nio.ByteBuffer;
 import java.nio.channels.SeekableByteChannel;
+import java.nio.file.AccessDeniedException;
 import java.nio.file.CopyOption;
 import java.nio.file.FileAlreadyExistsException;
 import java.nio.file.FileSystem;
@@ -351,6 +352,21 @@ public class FsBlobContainerTests extends ESTestCase {
             }
         }.getFileSystem(null);
         PathUtilsForTesting.installMock(noHardLinkFs);
+        try {
+            checkAtomicWrite();
+        } finally {
+            PathUtilsForTesting.installMock(fileSystem);
+        }
+    }
+
+    public void testAtomicWriteHardLinkThrowsAccessDeniedException() throws IOException {
+        final var noPermissionsForHardLinksFs = new FilterFileSystemProvider("nohardlinkfs://", fileSystem) {
+            @Override
+            public void createLink(Path link, Path existing) throws IOException {
+                throw new AccessDeniedException(link.toString(), existing.toString(), "no permissions for hard links");
+            }
+        }.getFileSystem(null);
+        PathUtilsForTesting.installMock(noPermissionsForHardLinksFs);
         try {
             checkAtomicWrite();
         } finally {

